### PR TITLE
refactor: save ydoc state in the database, so it can be restored easier

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
Realtime communication

### Description
By storing the ydoc state in the database we can reconnect lost clients easier and enable offline editing because we continue using the crdt data that has been used by the client before the connection loss.

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### CODE REVIEW

File: frontend/src/components/editor-page/editor-pane/hooks/yjs/use-code-mirror-yjs-extension.ts [hunk 1]
```diff
@@ -8,30 +8,33 @@ import { YTextSyncViewPlugin } from '../../codemirror-extensions/document-sync/y
...
 import type { YDocSyncClientAdapter } from '@hedgedoc/commons'
+import type { RealtimeDoc } from '@hedgedoc/commons'
 import { useEffect, useMemo, useState } from 'react'
-import type { Text as YText } from 'yjs'
 
 /**
  * Creates a {@link Extension code mirror extension} that synchronizes an editor with the given {@link YText ytext}.
  *
- * @param yText The source and target for the editor content
+ * @param doc The {@link RealtimeDoc realtime doc} that contains the markdown contentn text channel
```
Comment (ErikMichelson) [inline frontend/src/components/editor-page/editor-pane/hooks/yjs/use-code-mirror-yjs-extension.ts:17]:
```suggestion
 * @param doc The {@link RealtimeDoc realtime doc} that contains the markdown content text channel
```

File: backend/src/realtime/realtime-note/realtime-note.ts [hunk 2]
```diff
@@ -28,12 +28,13 @@ export class RealtimeNote extends EventEmitter2<RealtimeNoteEventMap> {
   private readonly clients = new Set<RealtimeConnection>();
   private isClosing = false;
 
-  constructor(private readonly note: Note, initialContent: string) {
+  constructor(private readonly note: Note, initialState: string | number[]) {
```
Comment (DerMolly) [inline backend/src/realtime/realtime-note/realtime-note.ts:31]: Maybe we could use multiple contructors for those two different cases?

File: commons/src/y-doc-sync/realtime-doc.ts [hunk 3]
```diff
@@ -3,27 +3,53 @@
...
+
   /**
    * Creates a new instance.
    *
    * The new instance is filled with the given initial content.
    *
    * @param initialContent - the initial content of the {@link Doc YDoc}
    */
-  constructor(initialContent?: string) {
+  constructor(initialContent?: string | number[]) {
```
Comment (DerMolly) [inline commons/src/y-doc-sync/realtime-doc.ts:39]: same here

File: commons/src/y-doc-sync/realtime-doc.ts [hunk 4]
```diff
@@ -3,27 +3,53 @@
...
     super()
     if (initialContent) {
-      this.getMarkdownContentChannel().insert(0, initialContent)
+      if (typeof initialContent === 'string') {
```
Comment (davidmehren) [inline commons/src/y-doc-sync/realtime-doc.ts:42]: I'm a bit unhappy with this union type and the following `typeof`.

Actually, with the whole concept of mixing "a string of initial content" with "an arbitrary binary blob of yjs state" that gets introduced with this PR. I think these two types of data should be more cleanly separated.

**Review by DerMolly — CHANGES_REQUESTED** [review]

File: backend/src/realtime/realtime-note/realtime-note.ts (hunk 2 above)
Comment (mrdrogdrog) [reply backend/src/realtime/realtime-note/realtime-note.ts:31]: Not possible in typescript. The way to go here is a union type.

Comment (ErikMichelson) [reply backend/src/realtime/realtime-note/realtime-note.ts:31]: You could add two separate static methods to the class as factory methods

**Review by davidmehren — APPROVED** [review]

**Review by DerMolly — APPROVED** [review]

**Review by ErikMichelson — APPROVED** [review]

**Pull request merged**